VoxelSniperCore/src/main/java/com/github/kevindagame/util/schematic/DataFolderSchematicReader.kt

```diff
@@ -0,0 +1,109 @@
+package com.github.kevindagame.util.schematic
+
+import com.github.kevindagame.VoxelSniper
+import net.sandrohc.schematic4j.SchematicLoader
+import net.sandrohc.schematic4j.schematic.Schematic
+import java.io.File
+
+class DataFolderSchematicReader : ISchematicReader {
+    override fun getSchematicFile(name: String): File? {
+        //check for exact name
+        val exactSchematic = VoxelSniper.voxelsniper.fileHandler.getDataFile(SCHEMATIC_FILE_ROOT_PATH + name)
+        if (exactSchematic.isFile) {
+            return exactSchematic
+        }
+
+        //if no, check for name with any extension
+        val schematicFiles = VoxelSniper.voxelsniper.fileHandler.getDataFile(SCHEMATIC_FILE_ROOT_PATH).listFiles()
+        if (schematicFiles != null) {
+            for (schematic in schematicFiles) {
+                if (schematic.isFile) {
+                    if (schematic.nameWithoutExtension == name) {
+                        return schematic
+                    }
+                }
+            }
+        }
+
+        return null
+    }
+
+    override fun getSchematicFolder(name: String): File? {
+        val path = SCHEMATIC_FILE_ROOT_PATH + name
+        val schematicFolder = VoxelSniper.voxelsniper.fileHandler.getDataFile(path)
+        if (schematicFolder.isDirectory) {
+            return schematicFolder
+        }
+        return null
+    }
+
+    override fun readSchematicFile(file: File): VoxelSchematic {
+        val schematic: Schematic
+        try {
+            schematic = SchematicLoader.load(file)
+        } catch (e: Exception) {
+            throw IllegalArgumentException("Schematic ${file.name} is not valid")
+        }
+
+        val voxelSchematicBuilder = VoxelSchematicBuilder()
+
+        voxelSchematicBuilder.name = file.nameWithoutExtension
+
+        for (y in 0 until schematic.height()) {
```

Comment:
It's possible to use the block iterator `schematic.blocks().forEach(blockAndPos -> ...)` here. Should be slightly more efficient, especially in formats with sub-regions like Litematica.

Reply:
I hadn't considered subregions, I suppose it wouldn't break the schematic per se, but it would cause issues with our default paste mode (which would overwrite any block with air)